syntax = "proto3";

service Raft {
  rpc RequestVote (VoteRequest) returns (VoteResponse);
  rpc AppendEntries (AppendEntriesRequest) returns (AppendEntriesResponse);
}

// Client service for submitting operations
service RaftClient {
  rpc SubmitOperation (ClientRequest) returns (ClientResponse);
}

message VoteRequest {
  int32 term = 1;
  string candidate_id = 2;
}

message VoteResponse {
  int32 term = 1;
  bool vote_granted = 2;
}

message LogEntry {
  int32 term = 1;
  string command = 2;
  int32 index = 3;
}

message AppendEntriesRequest {
  int32 term = 1;
  string leader_id = 2;
  repeated LogEntry entries = 3;
  int32 prev_log_index = 4;
  int32 prev_log_term = 5;
  int32 leader_commit = 6;
}

message AppendEntriesResponse {
  int32 term = 1;
  bool success = 2;
  int32 match_index = 3;
}

// Client request/response messages
message ClientRequest {
  string operation = 1;
  string client_id = 2;
}

message ClientResponse {
  bool success = 1;
  string message = 2;
  string leader_id = 3;
}